### PR TITLE
[PM-8965] Register for push notifications more aggressively

### PR DESCRIPTION
**Pull request by KatherineInCode**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-8965

## 📔 Objective

Previously, the app was only registering for push notifications on launch, which meant that it would only register for an active user at that time. This now registers more aggressively, by checking any time the user goes to the vault list. It will also register after a user has granted permission for push notifications. However, it will also only attempt to register once per day (defined as 86400 seconds), to match the MAUI app.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessor.swift [hunk 1]
```diff
@@ -234,6 +258,7 @@ extension VaultListProcessor {
                 do {
                     _ = try await services.notificationService
                         .requestAuthorization(options: [.alert, .sound, .badge])
+                    await self.registerForNotifications()
```
Comment (fedemkr) [inline BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessor.swift:261]: 🤔 Shouldn't we check the result from `requestAuthorization` and only if it's `true` call `await self.registerForNotifications()`? [Docs](https://developer.apple.com/documentation/usernotifications/unusernotificationcenter/requestauthorization(options:completionhandler:))

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessorTests.swift [hunk 2]
```diff
@@ -129,6 +135,71 @@ class VaultListProcessorTests: BitwardenTestCase { // swiftlint:disable:this typ
...
+        XCTAssertFalse(application.registerForRemoteNotificationsCalled)
+        XCTAssertEqual(
+            stateService.notificationsLastRegistrationDates["1"],
+            timeProvider.presentTime.addingTimeInterval(-86399)
+        )
+    }
+
+    /// `perform(_:)` with `appeared` registers the device for notifications
+    /// if the device attempted registration more than one day (that is, 86400 seconds) ago.
+    func test_perform_appeared_notificationRegistration_moreThanADay() async throws {
```
Comment (fedemkr) [inline BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessorTests.swift:180]: 🎨 If I'm not mistaken this would be to exactly a day. Can another test be added for more than a day? i.e. 86401 so we have the three possible cases for `>=`.

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessor.swift (hunk 1 above)
Comment (KatherineInCode) [reply BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessor.swift:261]: Good call; I missed that

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessorTests.swift (hunk 2 above)
Comment (KatherineInCode) [reply BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessorTests.swift:180]: I was mostly trying to test the boundary conditions, but I can add another one explicitly for 86401

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessor.swift (hunk 1 above)
Comment (matt-livefront) [reply BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessor.swift:261]: You can register for notifications without yet having authorization (or if the user denied authorization). The benefit of that is if the user enables notifications in settings, they will start receiving notifications immediately without having to reopen the app.

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessorTests.swift (hunk 2 above)
Comment (fedemkr) [reply BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessorTests.swift:180]: Great thanks! Just in case my reasoning behind this is because on a given `x >= n` condition, if we don't add a test for `x > n` then all the tests will keep passing if the condition would change to `x == n`.

Comment (KatherineInCode) [reply BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessorTests.swift:180]: It makes sense, though the tests would still pass if the condition changed to `x == n || x == n+1`. Figuring out the optimal set of tests to cover as much as possible without creating so many tests that execution takes forever is a tricky thing.

**Review by fedemkr — APPROVED** [review]
